Add models with exact and simplified NMDA dynamics according to Brunel and Wang (2001)

janskaar opened the pull request:
Update: Removed draft.

Adds iaf_wang_2002, a model with a conductance based exponential synapse with extra dynamics, from [1]. The model is simplified to allow efficient computation, by taking $t_\mathrm{AMPA,rise} \to 0$ and modelling it as a delta function. Since $t_\mathrm{AMPA,rise} \ll t_\mathrm{AMPA,decay}$, this should not make too much of a difference.


[1] Wang, X. J. (2002). Probabilistic decision making by slow reverberation in cortical circuits. Neuron, 36(5), 955-968.
       DOI: https://doi.org/10.1016/S0896-6273(02)01092-9

nicolossus reviewed (changes requested):
@janskaar Thanks! Here is my initial review after looking at the implementation of the model.

heplesser commented on testsuite/pytests/test_iaf_wang_2002.py:
```diff
@@ -0,0 +1,123 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Tests synaptic dynamics of iaf_wang_2002. Since the neuron is conductance based,
```
Put brief description on first line separated by blank line. Applies to all docstrings.

Make even more explicit that you test here against the approximation made, not the full model.

And one could in principle set up a test for models with untractable dynamics by using a SciPy ODE solver.

heplesser commented on testsuite/pytests/test_iaf_wang_2002.py:
```diff
@@ -0,0 +1,123 @@
+                                         "tau_GABA": tau_GABA,
+                                         "tau_decay_NMDA": tau_NMDA})
+    
+    pg = nest.Create("poisson_generator", {"rate": 50.})
+    sr = nest.Create("spike_recorder")
```
Why the devices between the neurons?

heplesser commented on testsuite/pytests/test_iaf_wang_2002.py:
```diff
@@ -0,0 +1,123 @@
+    
+    ex_syn_spec = {"synapse_model": "static_synapse",
+                    "weight": w_ex,
+                   "receptor_type": 0}
```
Do you need the receptor type argument? Also, `static_synpase` is default and need not be given explicitly.

heplesser commented on testsuite/pytests/test_iaf_wang_2002.py:
```diff
@@ -0,0 +1,123 @@
+    in_syn_spec = {"synapse_model": "static_synapse",
+                    "weight": w_in}
+    
+    conn_spec = {"rule": "all_to_all"}
```
`all_to_all` is default.

heplesser commented on testsuite/pytests/test_iaf_wang_2002.py:
```diff
@@ -0,0 +1,123 @@
+    nest.Connect(nrn1, nrn2, syn_spec=ex_syn_spec, conn_spec=conn_spec)
+    nest.Connect(nrn1, nrn2, syn_spec=in_syn_spec, conn_spec=conn_spec)
+    nest.Connect(mm1, nrn1)
+    
```
Why empty line here?

heplesser commented on testsuite/pytests/test_iaf_wang_2002.py:
```diff
@@ -0,0 +1,123 @@
+    nest.Simulate(1000.)
+    
+    # get spike times from membrane potential
+    # cannot use spike_recorder because we abuse exact spike timing
```
Yes you can: If you set `time_in_steps_` on the recorder, you get `time_steps` in the `event` dictionary, which just needs to be multiplied with the resolution to get the spike time in ms. In the `time_offset` field you even get the value that is sent along.

heplesser commented on pynest/examples/wang_neuron.py:
```diff
@@ -0,0 +1,101 @@
+import nest
```
This example lacks comments and the copyright header. The example should explain how to record spikes (see comment in test below).

heplesser commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,181 @@
+import nest
```
Lacks copyright header and all documentation. See other examples for style for comments that is required for our conversion to notebooks.

heplesser commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,181 @@
+               "tau_decay_NMDA": 100.0,
+               "alpha": 0.5,
+               "conc_Mg2": 1.0,
+               "g_L": 25.,               # leak conductance
```
Why comments only from here on? It would be good to give units.

heplesser commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,181 @@
+s_AMPA = mm_selective1.get("events", "s_AMPA")[inds]
+s_AMPA_ext = mm_selective1.get("events", "s_AMPA_ext")[inds]
+s_GABA = mm_selective1.get("events", "s_GABA")[inds]
+s_NMDA = mm_selective1.get("events", "s_NMDA")[inds]
```
More output, please :).

heplesser commented on nestkernel/nest_names.cpp:
```diff
@@ -106,6 +106,7 @@ const Name continuous( "continuous" );
 const Name count_covariance( "count_covariance" );
 const Name count_histogram( "count_histogram" );
 const Name covariance( "covariance" );
+const Name conc_Mg2( "conc_Mg2" );
```
I am not entirely happy with `conc`, but struggle to find something better.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+    double V_reset;        //!< Reset Potential in mV
+    double C_m;            //!< Membrane Capacitance in pF
+    double g_L;            //!< Leak Conductance in nS
+    double g_GABA;         //!< Peak conductance GABA
```
Give units for all.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+    {
+      V_m = 0,
+      s_AMPA,
+      s_AMPA_ext,
```
What is this field?

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+    State_( const Parameters_& ); //!< Default initialization
+    State_( const State_& );
+
+    double s_NMDA_pre = 0;
```
Comment.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+    // -----------------------------------------------------------------------
+    //   Buffers and sums of incoming spikes and currents per timestep
+    // -----------------------------------------------------------------------
+    RingBuffer spike_AMPA;
+    RingBuffer spike_AMPA_ext;
+    RingBuffer spike_GABA;
+    RingBuffer spike_NMDA;
```
For consistency, append `_` to names.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+inline size_t
+iaf_wang_2002::handles_test_event( SpikeEvent&, size_t receptor_type )
+{
+  if ( receptor_type == 0 )
+  {
+    return 0;
+  }
+  else if ( receptor_type == 1 )
+  {
+    return 1;
+  }
+  else
+  {
+    throw UnknownReceptorType( receptor_type, get_name() );
+  }
```
Use names from enum.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,512 @@
+  S_.get( d );
+  ArchivingNode::get_status( d );
+
+  DictionaryDatum receptor_type = new Dictionary();
+
+//  ( *d )[ names::receptor_types ] = receptor_type;
```
It would be good to have the `receptor_types` dictionary, so users can use them model without having to remember what is 1, 2.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+    INF_SPIKE_RECEPTOR = 0,
+    AMPA_NMDA,
+    GABA,
+    SUP_SPIKE_RECEPTOR
```
Not really used below, but shoud be.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,488 @@
+
+This model implements a version of the neuron model described in [1]_.
+
+It contains AMPA, GABA and NMDA synapses, where the number of NMDA ports are dependent
+on the number of presynaptic connections.
+
+The AMPA and GABA synapses are given as alpha functions, while the NMDA synapse is modeled
+with a non-linear function described by
+
+.. math::
+   \frac{ dg_j^{NMDA}(t) }{ dt } = - \frac{ g_j^{NMDA}(t) }{ \tau_{NMDA,decay} } + \alpha x_j(t)(1 - s_j^{NMDA}(t)) \\
+   \frac{ dx_j(t) }{ dt } =- \frac{ x_j(t) }{ \tau_{NMDA,rise} } + \sum_k \delta(t - t_j^k).
+
+The synaptic current of NMDA is given by
+
+.. math::
+   I_{NMDA}(t) = \frac{ V(t) - E_{ex} }{ 1 + [Mg^{2+}]exp(-0.062V(t))/3.57 }\sum_{j=1}w_jg_j^{NMDA},
+
+where `w_j` is the weight of connection with presynaptic neuron `j`.
+
```
I think this is not up to date.

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,541 @@
+    S_.y_[ State_::s_AMPA_ext ] += B_.spike_AMPA_ext.get_value( lag );
+    S_.y_[ State_::s_GABA ] += B_.spike_GABA.get_value( lag );
+    S_.y_[ State_::s_NMDA ] += B_.spike_NMDA.get_value( lag );
+    if ( S_.y_[ State_::s_NMDA ] > 1 )
+    {
+      S_.y_[ State_::s_NMDA ] = 1;
+    }
```
```suggestion
    S_.y_[ State_::s_NMDA ] = std::min( S_.y_[ State_::s_NMDA ], 1 );
```

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,541 @@
+
+      // compute current value of s_NMDA and add NMDA update to spike offset
+      S_.s_NMDA_pre = S_.s_NMDA_pre * exp( -( t_spike - t_lastspike ) / P_.tau_decay_NMDA );
+      double s_NMDA_delta = P_.alpha * (1 - S_.s_NMDA_pre);
```
```suggestion
      const double s_NMDA_delta = P_.alpha * (1 - S_.s_NMDA_pre);
```

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,541 @@
+      B_.spike_NMDA.add_value( e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ),
+        e.get_weight() * e.get_multiplicity() * e.get_offset() );
+    }
+    else if ( e.get_rport() == 1 ) { // external spike
+      B_.spike_AMPA_ext.add_value( e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ),
+        e.get_weight() * e.get_multiplicity() );
```
Why this, what is an "external spike"?

Since offset is 0 for spikes sent from other neuron models (except precise spiking, might need a check here), we could run all spikes through rport 0.

jessica-mitchell commented on pynest/examples/wang_decision_making.py:
@janskaar  Can you please update the PyNEST examples to include a docstring at the beginning (see how other examples are set up) so sphinx-gallery can render the example correctly; right now Read the Docs fails to build the docs entirely. Also it's necessary to explain what this example is for. The links to  examples also have to be included into doc/htmldoc/examples/index.rst,so they can be discoverable.

heplesser commented:
@janskaar Several formatting test fail. Could you get them fixed?

janskaar commented:
> @janskaar Several formatting test fail. Could you get them fixed?

All the checks pass now. I have implemented all the changes we discussed, apart from one. I still have both num_ports and state_vec_size variables, as I think the code is a lot easier to understand with separate variables instead of having to do  more arithmetic.

I also added the notebook with a comparison between exact/approximate model and the equations in model_details.

heplesser commented:
@janskaar Thanks for the updates, but on the CI tests fail now when GSL is available, i.e., the model included. Could you check what is going on? I'd also suggest to merge master since that removes the touchy mini_brunel_ps MPI test.

heplesser commented on models/iaf_wang_2002_exact.cpp:
```diff
@@ -0,0 +1,568 @@
+    const size_t w_idx = rport - NMDA;
+    if ( B_.weights_[ w_idx ] != e.get_weight() )
+    {
+      B_.weights_[ w_idx ] = e.get_weight();
```
Throw exception because weight change cannot be implemented properly due to way model is defined.

janskaar replied:
Added

janskaar commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,522 @@
+  // helper vars
+  const double alpha_tau = P_.alpha * P_.tau_rise_NMDA;
+  const double tau_rise_tau_dec = P_.tau_rise_NMDA / P_.tau_decay_NMDA;
+  const double expint = boost::math::gamma_q( 1 - tau_rise_tau_dec, alpha_tau )
+    * boost::math::tgamma( 1 - tau_rise_tau_dec ) * pow( alpha_tau, tau_rise_tau_dec - 1 );
+
+
+  V_.S_jump_1 = exp( -P_.alpha * P_.tau_rise_NMDA ) - 1;
+  V_.S_jump_0 = -expint * alpha_tau + pow( alpha_tau, tau_rise_tau_dec ) * boost::math::tgamma( 1 - tau_rise_tau_dec );
```
It's the wrong expint unfortunately. The upper incomplete gamma function we use now is not there either, so we have to continue using boost.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,518 @@
+The membrane potential and synaptic variables evolve according to
+
+.. math::
+
```
Some of the math code here seems off so it is not rendered. You can see the rendered result at https://nest-simulator--2979.org.readthedocs.build/en/2979/models/iaf_wang_2002.html

Building the docs on your computer should also emit LaTeX error messages.

heplesser commented on models/iaf_wang_2002_exact.h:
```diff
@@ -0,0 +1,516 @@
+    \Gamma_\mathrm{ex}}^{N_E}w_jS_{j,\mathrm{NMDA}}(t) \\[3ex]
+    I_\mathrm{GABA} &= (V(t) - V_I)\sum_{j \in \Gamma_\mathrm{in}}^{N_E}w_jS_{j,\mathrm{GABA}}(t) \\[5ex]
+    \frac{dS_{j,\mathrm{AMPA}}}{dt} &=-\frac{j,S_{\mathrm{AMPA}}}{\tau_\mathrm{AMPA}}+\sum_{k \in \Delta_j} \delta (t -
+t_j^k) \\[3ex] \frac{dS_{j,\mathrm{GABA}}}{dt} &= -\frac{S_{j,\mathrm{GABA}}}{\tau_\mathrm{GABA}} + \sum_{k \in
```
The LaTeX does not render fully, see https://nest-simulator--2979.org.readthedocs.build/en/2979/models/iaf_wang_2002_exact.html

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,518 @@
+
+The following parameters can be set in the status dictionary.
+
+
```
```suggestion
```

Could you format the table as in https://nest-simulator.readthedocs.io/en/stable/models/iaf_psc_alpha.html ?

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,518 @@
+
+.. note::
+   :math:`g_{\mathrm{\{\{rec,AMPA\}, \{ext,AMPA\}, GABA, NMBA}\}}` from [1]_ is built into the weights in this NEST
+model, so setting the variables is thus done by changing the weights.
```
```suggestion
model, so these variables are set by changing the weights.
```

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,518 @@
+
+.. note::
+    For the NMDA dynamics to work, the both pre-synaptic and post-synaptic neuron must be of type iaf_wang_2002. For
+AMPA/GABA synapses, any pre-synaptic neuron can be used.
```
Could you also add a note like

For technical reasons, spikes from `iaf_wang_2002` neurons must be recorded with `time_in_steps: True` set in the spike recorder, ignoring the offset value. We hope to correct this in a future version of NEST.

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,518 @@
+Examples using this model
++++++++++++++++++++++++++
+
+.. listexamples:: ht_neuron
```
I don't think so, but we have an better one, don't we?

heplesser commented on models/iaf_wang_2002.h:
```diff
@@ -0,0 +1,518 @@
+
+.. listexamples:: ht_neuron
+
+
+
```
```suggestion
```

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,521 @@
+  // not the state vector in the node, node.S_.y[].
+
+  const double I_AMPA = ( y[ S::V_m ] - node.P_.E_ex ) * y[ S::s_AMPA ];
+
```
```suggestion
```

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,521 @@
+  // helper vars
+  const double alpha_tau = P_.alpha * P_.tau_rise_NMDA;
+  const double tau_rise_tau_dec = P_.tau_rise_NMDA / P_.tau_decay_NMDA;
+  const double expint = boost::math::gamma_q( 1 - tau_rise_tau_dec, alpha_tau )
```
The function names here do not match the ones in the notebook. Provide some explanation.

janskaar replied:
This has now been updated with a simpler expression in both the notebook and in the code

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,521 @@
+      }
+    }
+
+
```
```suggestion
```

heplesser commented on models/iaf_wang_2002_exact.h:
```diff
@@ -0,0 +1,516 @@
+
+The following parameters can be set in the status dictionary.
+
+
```
```suggestion
```

Also here, reformat table as in iaf_psc_alpha.

heplesser commented on models/iaf_wang_2002_exact.h:
```diff
@@ -0,0 +1,516 @@
+The specification of this model differs slightly from the one in [1]_. The parameters :math:`g_\mathrm{AMPA}`,
+:math:`g_\mathrm{GABA}`, and :math:`g_\mathrm{NMDA}` have been absorbed into the respective synaptic weights.
+Additionally, the synapses from the external population is not separated from the recurrent AMPA-synapses.
+
```
Should point here to `iaf_wang_2002` as the more efficient variant.

heplesser commented on models/iaf_wang_2002_exact.h:
```diff
@@ -0,0 +1,516 @@
+See also
+++++++++
+
+iaf_wang_2002
```
No examples?

heplesser commented on models/iaf_wang_2002_exact.cpp:
```diff
@@ -0,0 +1,570 @@
+  // not the state vector in the node, node.S_.ode_state[].
+
+  const double I_AMPA = ( ode_state[ State_::V_m ] - node.P_.E_ex ) * ode_state[ State_::s_AMPA ];
+
```
```suggestion
```

heplesser commented on models/iaf_wang_2002_exact.cpp:
```diff
@@ -0,0 +1,570 @@
+  assert( pnode );
+  nest::iaf_wang_2002_exact& node = *( reinterpret_cast< nest::iaf_wang_2002_exact* >( pnode ) );
+
+  // ode_state[] here is---and must be---the state vector supplied by the integrator,
```
Why must it be? Explain.

janskaar replied:
These comments are from the iaf_cond_exp model I copied, and are identical in all iaf_cond_* models. If we want to change it, would it be better to do it for all models in a separate PR?

heplesser replied:
OK, then let's leave it.

heplesser commented on models/iaf_wang_2002_exact.cpp:
```diff
@@ -0,0 +1,570 @@
+    for ( size_t i = NMDA - 1; i < B_.spikes_.size(); ++i )
+    // i starts at 2, runs through all NMDA spikes
+    {
+
+      // index which starts at 0
+      const size_t si = i - ( NMDA - 1 );
```
```suggestion
      const size_t si = i - ( NMDA - 1 );        // index which starts at 0
```

heplesser commented on models/iaf_wang_2002_exact.h:
```diff
@@ -0,0 +1,516 @@
+    if ( receptor_type == SynapseTypes::NMDA )
+    {
+      // give each NMDA synapse a unique rport, starting from 3 (num_ports_ is initialized to 3)
+      ++S_.num_ports_;
```
After we have run `init_buffers_()`, i.e., once `B_.e_ != nullptr`, the size of the state vector is fixed and we cannot add any new NMDA connections. We need a check for that here.

janskaar replied:
I added a check for this, see the new version

heplesser reviewed (changes requested):
@janskaar Nice progress. I have seen through the f`iaf_wang_2002*.{h,cpp}` files and added some comments. I will review the tests and examples later.

heplesser commented on models/iaf_wang_2002.cpp:
```diff
@@ -0,0 +1,516 @@
+  const double alpha_tau = P_.alpha * P_.tau_rise_NMDA;
+  const double tau_rise_tau_dec = P_.tau_rise_NMDA / P_.tau_decay_NMDA;
+
+  V_.S_jump_1 = exp( -P_.alpha * P_.tau_rise_NMDA ) - 1;
+  V_.S_jump_0 = pow( alpha_tau, tau_rise_tau_dec ) * boost::math::tgamma_lower( 1 - tau_rise_tau_dec, alpha_tau );
```
NEST guidelines suggest to always use `std::` and `expm1()` is more precise.

```suggestion
  V_.S_jump_1 = std::expm1( -P_.alpha * P_.tau_rise_NMDA );
  V_.S_jump_0 = std::pow( alpha_tau, tau_rise_tau_dec ) * boost::math::tgamma_lower( 1 - tau_rise_tau_dec, alpha_tau );
```

heplesser reviewed (changes requested):
Just one little suggestion now, still need to look at examples and tests.

Also consider to rename all _jump variables to _post, see notebook.

jessica-mitchell commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,424 @@
+    # the signal is given by a time-inhomogeneous Poisson process,
+    # where the expectations are constant over intervals of 50ms,
+    # and then change. The values for each interval are normally
+    # distributed, with means mu_a and mu_b, and standard deviation
```
```suggestion
    # distributed, with means `mu_a`and `mu_b`, and standard deviation
```

jessica-mitchell commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,424 @@
+    # where the expectations are constant over intervals of 50ms,
+    # and then change. The values for each interval are normally
+    # distributed, with means mu_a and mu_b, and standard deviation
+    # sigma.
```
```suggestion
    # `sigma`.
```

jessica-mitchell commented on models/iaf_bw_2001_exact.h:
```diff
@@ -0,0 +1,529 @@
+Short description
++++++++++++++++++
+
+Leaky integrate-and-fire-neuron model with conductance based synapses, and additional NMDA receptors.
```
```suggestion
Leaky integrate-and-fire-neuron model with conductance based synapses and additional NMDA receptors.
```

jessica-mitchell commented on models/iaf_bw_2001_exact.h:
```diff
@@ -0,0 +1,529 @@
+Since :math:`S_{j,\mathrm{AMPA}}` and :math:`S_{j,\mathrm{GABA}}` are piecewise exponential functions, the sums are also
+a piecewise exponential function, and can be stored in a single synaptic variable each, :math:`S_{\mathrm{AMPA}}` and
+:math:`S_{\mathrm{GABA}}` respectively. The sum over :math:`S_{j,\mathrm{NMDA}}` does not have a simple expression, and
+cannot be simplified. Therefore, for each synapse, we need to integrate separate state variable, which makes the model
```
```suggestion
cannot be simplified. Therefore, for each synapse, we need to integrate separate state variables, which makes the model
```
Not sure if variables should be plural or missing an "a" before _separate_ if singular

jessica-mitchell commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+Short description
++++++++++++++++++
+
+Leaky integrate-and-fire-neuron model with conductance based synapses, and additional NMDA receptors.
```
```suggestion
Leaky integrate-and-fire-neuron model with conductance based synapses and additional NMDA receptors.
```

jessica-mitchell commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+    k_1 &= \mathrm{exp}(-\alpha \tau_\mathrm{r}) - 1
+
+where :math:`\gamma` is the `lower incomplete gamma function
+<https://en.wikipedia.org/wiki/Incomplete_gamma_function>`_. For these values of :math:`k_0` and :math:`k_1`, the approximate model will approach the exact model for large t.
```
```suggestion
<https://en.wikipedia.org/wiki/Incomplete_gamma_function>`_. For these values of :math:`k_0` and :math:`k_1`, the approximate model will approach the exact model for large `t`.
```

jessica-mitchell commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+
+The specification of this model differs slightly from the one in [1]_. The parameters :math:`g_\mathrm{AMPA}`,
+:math:`g_\mathrm{GABA}`, and :math:`g_\mathrm{NMDA}` have been absorbed into the respective synaptic weights.
+Additionally, the synapses from the external population is not separated from the recurrent AMPA-synapses.
```
```suggestion
Additionally, the synapses from the external population are not separated from the recurrent AMPA-synapses.
```

jessica-mitchell commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+================== ================= ========================== =================================
+
+.. note::
+   :math:`g_{\mathrm{\{\{rec,AMPA\}, \{ext,AMPA\}, GABA, NMBA}\}}` from [1]_ is built into the weights in this NEST model, so these variables are set by changing the weights.
```
```suggestion
   :math:`g_{\mathrm{\{\{rec,AMPA\}, \{ext,AMPA\}, GABA, NMBA}\}}` from [1]_ are built into the weights in this NEST model, so these variables are set by changing the weights.
```

jessica-mitchell commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+   :math:`g_{\mathrm{\{\{rec,AMPA\}, \{ext,AMPA\}, GABA, NMBA}\}}` from [1]_ is built into the weights in this NEST model, so these variables are set by changing the weights.
+
+.. note::
+    For the NMDA dynamics to work, the both pre-synaptic and post-synaptic neuron must be of type iaf_bw_2001. For AMPA/GABA synapses, any pre-synaptic neuron can be used.
```
```suggestion
    For the NMDA dynamics to work,  both pre-synaptic and post-synaptic neurons must be of type ``iaf_bw_2001``. For AMPA/GABA synapses, any pre-synaptic neuron can be used.
```

jessica-mitchell commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+    For the NMDA dynamics to work, the both pre-synaptic and post-synaptic neuron must be of type iaf_bw_2001. For AMPA/GABA synapses, any pre-synaptic neuron can be used.
+
+.. note::
+    For technical reasons, spikes from iaf_bw_2001 neurons must be recorded with time_in_steps: True set in the spike recorder, ignoring the offset value. We hope to correct this in a future version of NEST.
```
```suggestion
    For technical reasons, spikes from ``iaf_bw_2001`` neurons must be recorded with ``time_in_steps: True`` set in the spike recorder, ignoring the offset value. We hope to correct this in a future version of NEST.
```

jessica-mitchell commented on models/iaf_bw_2001_exact.h:
```diff
@@ -0,0 +1,529 @@
+Examples using this model
++++++++++++++++++++++++++
+
+.. listexamples:: iaf_bw_2001
```
```suggestion
.. listexamples:: iaf_bw_2001_exact
```

jessica-mitchell reviewed (changes requested):
Thanks @janskaar I have a few suggested changes.

Also, we need to add this example to the [example index](https://nest-simulator.readthedocs.io/en/latest/examples/index.html) page; I guess this should be its own card unless do you or @heplesser think this fits with any other example?

janskaar commented on models/iaf_bw_2001.h:
```diff
@@ -0,0 +1,505 @@
+
+For more implementation details and a comparison to the exact version see:
+
+- `wong_approximate_implementation <../model_details/wong_approximate_implementation.ipynb>`_
```
Fixed

janskaar commented on models/iaf_bw_2001_exact.h:
```diff
@@ -0,0 +1,529 @@
+++++++++++
+
+.. [1] Wang, X.-J. (1999). Synaptic Basis of Cortical Persistent Activity: The Importance of NMDA Receptors to Working Memory. Journal of Neuroscience, 19(21), 9587–9603. https://doi.org/10.1523/JNEUROSCI.19-21-09587.1999
+.. [2] Wang, X.-J. (1999). Synaptic Basis of Cortical Persistent Activity: The Importance of NMDA Receptors to Working Memory. Journal of Neuroscience, 19(21), 9587–9603. https://doi.org/10.1523/JNEUROSCI.19-21-09587.1999
```
I added the correct citation now. They are all referenced in one of the first lines in the documentation.

jessica-mitchell reviewed (approved):
thanks! lgtm

jessica-mitchell commented:
@janskaar The example still needs to be added to the examples/index file. I created one but I cannot find your username is the list for creating PRs

janskaar commented:
> @janskaar The example still needs to be added to the examples/index file. I created one but I cannot find your username is the list for creating PRs

Perfect, thanks, I pulled it directly from your fork now.

heplesser reviewed (changes requested):
@janskaar I have finally reviewed example and tests. In the example, I added a suggestion to avoid mixing NEST and NumPy RNGs. In the tests, I mostly have smaller comments.

janskaar commented on testsuite/pytests/test_iaf_bw_2001_exact.py:
```diff
@@ -0,0 +1,176 @@
+    ampa_soln = spiketrain_response(times, wang_params["tau_AMPA"], spikes, w_ex)
+    gaba_soln = spiketrain_response(times, wang_params["tau_GABA"], spikes, np.abs(w_in))
+
+    nptest.assert_array_equal(mm3.events["V_m"], mm4.events["V_m"])
```
Fixed

janskaar commented on testsuite/pytests/test_iaf_bw_2001_exact.py:
```diff
@@ -0,0 +1,176 @@
+        conc_Mg2=1.0,  # Magnesium concentration
+    )
+
+    nrn1 = nest.Create("iaf_bw_2001_exact", wang_params)
```
Fixed

janskaar commented on testsuite/pytests/test_iaf_bw_2001.py:
```diff
@@ -0,0 +1,239 @@
+    nest.Connect(mm3, nrn3)
+
+    nest.Simulate(500.0)
+    assert np.max(np.abs(mm2.events["V_m"] - mm3.events["V_m"])) < 0.25
```
Fixed

heplesser reviewed (approved):
@janskaar Thanks, looks good now!

nicolossus commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,437 @@
+    update_times = np.arange(0, signal_duration, signal_update_interval)
+    update_times[0] = 0.1
+
+    # for p ... below allows us to create a local variable in the iterator which is created once
```
This comment looks a bit uncomplete

heplesser replied:
Your are right, here an improved version:
```suggestion
    # We could have written the generator expressions passed to `np.fromiter()` below as
    #
     #    ( nest.CreateParameter(...).GetValue() for _ in range(num_updates) )
     #
     # but that would create a new Parameter object for each iteration. We avoid this by
     # wrapping the iterator in an outer loop in which `p` assumes only a single value.
```

nicolossus commented on pynest/examples/wang_decision_making.py:
```diff
@@ -0,0 +1,437 @@
+
+    ##################################################
+    # Collect data from simulation
+    spikes_nonselective = sr_nonselective.get("events", "times")
+    spikes_selective1 = sr_selective1.get("events", "times")
+    spikes_selective2 = sr_selective2.get("events", "times")
+    spikes_inhibitory = sr_inhibitory.get("events", "times")
```
```suggestion
    spikes_nonselective = sr_nonselective.events["times"]
    spikes_selective1 = sr_selective1.events["times"]
    spikes_selective2 = sr_selective2.events["times"]
    spikes_inhibitory = sr_inhibitory.events["times"]
```

nicolossus commented on testsuite/pytests/test_iaf_bw_2001.py:
```diff
@@ -0,0 +1,234 @@
+
+    nest.Simulate(1000.0)
+
+    spikes = sr.get("events", "times") * nest.resolution
```
```suggestion
    spikes = sr.events["times"] * nest.resolution
```

nicolossus commented on testsuite/pytests/test_iaf_bw_2001_exact.py:
```diff
@@ -0,0 +1,173 @@
+
+    nest.Simulate(1000.0)
+
+    spikes = sr.get("events", "times") * nest.resolution
```
```suggestion
    spikes = sr.events["times"] * nest.resolution
```

nicolossus reviewed (approved):
Impressive work! I only have a few non-critical comments that you may consider, so I'll go ahead and approve 👍

janskaar commented on models/iaf_bw_2001_exact.h:
```diff
@@ -0,0 +1,544 @@
+// clang-format off
+/* BeginUserDocs: neuron, integrate-and-fire, conductance-based
+
+Short description
```
Changed the iaf_bw_2001 one slightly
